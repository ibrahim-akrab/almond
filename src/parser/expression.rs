//! Parsing for JS expressions.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::{branch::alt, bytes::complete::*, combinator::*, IResult};
use nom_locate::position;

/// Alias for `parse_expr_bp(s, 0)`.
pub fn parse_expr(s: Span) -> IResult<Span, Node> {
    parse_expr_bp(s, 0)
}

/// Parse an atomic expression â€” either a single token that is an
// expression, an expression started by a keyword like `function` or
// `new`, or an expression wrapped in punctuation like `()`, `[]`,
// or `{}`.
pub fn parse_primary_expr(s: Span) -> ParseResult<Node> {
    alt((
        parse_this_expr,
        parse_identifier,
        literal::parse_literal,
        parse_paren_expr,
    ))(s)
}

pub fn parse_this_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, _) = tag("this")(s)?;
    let (s, end) = position(s)?;
    let (s, _) = sp0(s)?;
    Ok((s, NodeKind::ThisExpression.with_pos(start, end)))
}

pub fn parse_paren_expr(s: Span) -> ParseResult<Node> {
    delimited(ws0(char('(')), parse_expr, ws0(char(')')))(s)
}

fn parse_opt_expr_in_list(s: Span) -> ParseResult<Option<Node>> {
    alt((
        value(None, peek(char(','))),
        map(parse_expr, |expr| Some(expr)),
    ))(s)
}

pub fn parse_expr_list_with_opt_expr(s: Span) -> ParseResult<Vec<Option<Node>>> {
    terminated(
        separated_list0(ws0(char(',')), parse_opt_expr_in_list),
        // eat trailing comma
        ws0(opt(char(','))),
    )(s)
}

pub fn parse_expr_list(s: Span) -> ParseResult<Vec<Node>> {
    terminated(
        separated_list0(ws0(char(',')), parse_expr),
        // eat trailing comma
        ws0(opt(char(','))),
    )(s)
}

fn parse_computed_member_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (object, property)) = tuple((
        parse_member_expr,
        delimited(ws0(char('[')), parse_expr, char(']')),
    ))(s)?;
    let (s, end): (Span, Span) = position(s)?;
    let (s, _) = sp0(s)?;
    Ok((
        s,
        NodeKind::MemberExpression {
            computed: true,
            property: Box::new(property),
            object: Box::new(object),
        }
        .with_pos(start, end),
    ))
}

fn parse_prop_member_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (object, property)) = tuple((
        parse_member_expr,
        preceded(ws0(char('.')), parse_identifier),
    ))(s)?;
    let (s, end): (Span, Span) = position(s)?;
    let (s, _) = sp0(s)?;
    Ok((
        s,
        NodeKind::MemberExpression {
            computed: false,
            property: Box::new(property),
            object: Box::new(object),
        }
        .with_pos(start, end),
    ))
}

fn parse_new_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (callee, arguments)) = preceded(
        ws1(keyword_new),
        pair(
            parse_member_expr,
            delimited(ws0(char('(')), parse_expr_list, char(')')),
        ),
    )(s)?;
    let (s, end) = position(s)?;
    let (s, _) = sp0(s)?;
    Ok((
        s,
        NodeKind::NewExpression {
            callee: Box::new(callee),
            arguments,
        }
        .with_pos(start, end),
    ))
}

pub fn parse_member_expr(s: Span) -> ParseResult<Node> {
    alt((
        // terminated(parse_primary_expr, not(alt((char('.'), char('['))))),
        parse_prop_member_expr,     // '.'
        parse_computed_member_expr, // '['
        parse_new_expr,             // new
        parse_primary_expr,
    ))(s)
}

pub fn parse_call_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (callee, arguments)) = pair(
        parse_member_expr,
        delimited(ws0(char('(')), parse_expr_list, char(')')),
    )(s)?;
    let (s, end) = position(s)?;
    let (s, _) = sp0(s)?;
    Ok((
        s,
        NodeKind::CallExpression {
            callee: Box::new(callee),
            arguments,
        }
        .with_pos(start, end),
    ))
}

pub fn parse_lhs_expr(s: Span) -> ParseResult<Node> {
    alt((parse_call_expr, parse_member_expr))(s)
}

// fn prefix_expr(s: Span) -> ParseResult<Node> {

// }

// fn postfix_expr(s: Span) -> ParseResult<Node> {

// }

// pub fn parse_unary_expr(s: Span) -> ParseResult<Node> {
//     ws0(alt((prefix_expr, postfix_expr)))(s)
// }

/// Pratt parsing for prefix operators. Called in `parse_expr_bp`.
fn parse_prefix_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (unary_op, BindingPower(_, right_bp))) = parse_prefix_operator(s)?;
    let (s, rhs) = parse_expr_bp(s, right_bp)?;
    let (s, end) = position(s)?;

    let node_kind = match unary_op {
        PrefixOperator::Unary(unary_op) => NodeKind::UnaryExpression {
            argument: Box::new(rhs),
            operator: unary_op,
            prefix: true,
        },
        PrefixOperator::Update(unary_op) => NodeKind::UpdateExpression {
            argument: Box::new(rhs),
            operator: unary_op,
            prefix: true,
        },
    };
    Ok((s, node_kind.with_pos(start, end)))
}

/// Pratt parsing for expressions with operator precedence.
/// Check out [https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html](https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html) to see how Pratt parsing works.
pub fn parse_expr_bp(s: Span, min_bp: i32) -> ParseResult<Node> {
    let (mut s, mut lhs) = alt((parse_prefix_expr, parse_primary_expr))(s)?;

    loop {
        // do not override s just yet
        let (s_tmp, (op, BindingPower(left_bp, right_bp))) = match parse_infix_operator(s) {
            Ok(res) => res,
            Err(_) => break, // do not return from function, just break from loop.
        };

        if left_bp < min_bp {
            break;
        }

        // ok, now we can override s
        s = s_tmp;

        let (s_tmp, rhs) = parse_expr_bp(s, right_bp)?;
        s = s_tmp;

        let start = lhs.clone().start;
        let end = rhs.clone().end;

        let node_kind = match op {
            InfixOperator::Binary(op) => NodeKind::BinaryExpression {
                left: Box::new(lhs),
                right: Box::new(rhs),
                operator: op,
            },
            InfixOperator::Logical(op) => NodeKind::LogicalExpression {
                left: Box::new(lhs),
                right: Box::new(rhs),
                operator: op,
            },
            InfixOperator::Assignment(op) => NodeKind::AssignmentExpression {
                left: Box::new(lhs),
                right: Box::new(rhs),
                operator: op,
            },
            InfixOperator::DotOperator => NodeKind::MemberExpression {
                object: Box::new(lhs),
                property: Box::new(rhs),
                computed: false,
            },
        };
        lhs = node_kind.with_pos(start, end);
    }

    Ok((s, lhs))
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn smoke_test_this_expr() {
        parse_this_expr("this".into()).unwrap();
        parse_this_expr("notthis".into()).unwrap_err();
    }

    #[test]
    fn smoke_test_primary_expr() {
        parse_primary_expr("this".into()).unwrap();
        parse_primary_expr("myVar".into()).unwrap();
    }

    #[test]
    fn smoke_test_paren_expr() {
        parse_paren_expr("(true)".into()).unwrap();
    }

    #[test]
    #[ignore]
    fn test_member_expr() {
        assert_json_snapshot!(parse_member_expr("a.b".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("a.b.c".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("a[1]".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("a[0]".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("a[[]]".into()).unwrap().1);
    }

    #[test]
    #[ignore]
    fn test_new_expr() {
        assert_json_snapshot!(parse_member_expr("new Array()".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("new Array(1)".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("new Array(1,)".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("new Array(1,2)".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("new Foo.Bar(true)".into()).unwrap().1);
    }

    #[test]
    #[ignore]
    fn test_call_expr() {
        assert_json_snapshot!(parse_call_expr("foo()".into()).unwrap().1);
        assert_json_snapshot!(parse_call_expr("foo.bar()".into()).unwrap().1);
        assert_json_snapshot!(parse_call_expr("foo.bar.baz()".into()).unwrap().1);
        assert_json_snapshot!(parse_call_expr("foo.bar(baz)".into()).unwrap().1);
        assert_json_snapshot!(parse_call_expr("foo.bar(baz, 1, 2, 3)".into()).unwrap().1);
    }

    #[test]
    fn test_expr_bp_infix() {
        assert_json_snapshot!(parse_expr("1 + 2".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 - 2 - 3".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 * 2 + 3".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 + 2 * 3".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 * 2 + 3 * 4".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("(1 + 2) * 3".into()).unwrap().1);
    }
    #[test]
    fn test_expr_bp_prefix() {
        assert_json_snapshot!(parse_expr("-1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("+1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("+(+1)".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 + -2".into()).unwrap().1); // same as 1 + (-1)

        assert_json_snapshot!(parse_expr("++x".into()).unwrap().1);
    }

    #[test]
    fn test_expr_bp_assignment() {
        assert_json_snapshot!(parse_expr("x = 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x = 1 + 2".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x += 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x += y += 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x += x * x".into()).unwrap().1);
    }

    #[test]
    fn test_expr_bp_member_expr() {
        assert_json_snapshot!(parse_expr("x.y".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x.y.z".into()).unwrap().1);
    }
}
