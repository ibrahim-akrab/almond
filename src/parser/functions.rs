//! Parsing for JS functions and programs.

