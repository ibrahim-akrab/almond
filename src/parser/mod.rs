//! Parsing for JS
//! # Whitespace Handling
//! All functions named `parse_*` should handle leading whitespace. Preceding whitespace is only handled in top level parse function.

mod expression;
mod functions;
mod identifier;
mod keyword;
mod literal;
mod precedence;
mod regex;
mod statement;
mod util;
pub use expression::*;
pub use functions::*;
pub use identifier::*;
pub use keyword::*;
pub use literal::*;
pub use precedence::*;
pub use regex::*;
pub use statement::*;
pub use util::*;

use nom::{
    branch::alt, bytes::complete::*, character::complete::*, combinator::*, error::context,
    multi::*, number::complete::*, sequence::*,
};
