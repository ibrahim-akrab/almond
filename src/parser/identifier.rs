use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::{
    branch::alt, bytes::complete::*, character::complete::*, character::*, combinator::*,
    number::complete::*, IResult,
};
use nom_locate::position;

/// Parses any 4 hex digits unicode escape sequence
fn parse_unicode_escape_sequence(s: Span) -> IResult<Span, char> {
    let (s, _) = tag("\\u")(s)?;
    let (s, hex_str) = recognize(count(one_of("1234567890abcdefABCDEF"), 4))(s)?;
    let hex_u32 = u32::from_str_radix(*hex_str, 16).unwrap(); // FIXME
    let char = std::char::from_u32(hex_u32).unwrap(); // FIXME
    Ok((s, char))
}

pub fn parse_identifier(s: Span) -> IResult<Span, Node> {
    fn parse_identifier_start(s: Span) -> IResult<Span, char> {
        let (s, c) = verify(
            alt((parse_unicode_escape_sequence, none_of(""))),
            |c: &char| match c {
                c if unicode_xid::UnicodeXID::is_xid_start(*c) => true,
                '$' | '_' => true,
                _ => false,
            },
        )(s)?;
        Ok((s, c))
    }
    fn parse_identifier_continue(s: Span) -> IResult<Span, char> {
        let (s, c) = verify(
            alt((parse_unicode_escape_sequence, none_of(""))),
            |c: &char| match c {
                c if unicode_xid::UnicodeXID::is_xid_continue(*c) => true,
                '$' | '_' => true,
                _ => false,
            },
        )(s)?;
        Ok((s, c))
    }

    let (s, start) = position(s)?;
    let (s, name) = map(
        pair(parse_identifier_start, many0(parse_identifier_continue)),
        |(c, v)| format!("{}{}", c, v.into_iter().collect::<String>()),
    )(s)?;
    let (s, end) = position(s)?;

    Ok((s, NodeKind::Identifier { name }.with_pos(start, end)))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_identifier() {
        all_consuming(parse_identifier)("myVar".into()).unwrap();
        all_consuming(parse_identifier)("abc123".into()).unwrap();
        all_consuming(parse_identifier)("$".into()).unwrap();
        all_consuming(parse_identifier)("_".into()).unwrap();
        all_consuming(parse_identifier)("$elem".into()).unwrap();
        all_consuming(parse_identifier)("_elem".into()).unwrap();
        all_consuming(parse_identifier)("$123".into()).unwrap();
        all_consuming(parse_identifier)("_123".into()).unwrap();

        // bad identifier
        all_consuming(parse_identifier)("123abc".into()).unwrap_err();
    }

    #[test]
    fn test_identifier_with_unicode_escape() {
        assert_eq!(
            parse_identifier(r#"abc\u0061"#.into()).unwrap().1.kind,
            NodeKind::Identifier {
                name: "abca".into()
            }
        );
        assert_eq!(
            parse_identifier(r#"abc\u0061\u0062123"#.into()).unwrap().1.kind,
            NodeKind::Identifier {
                name: "abcab123".into()
            }
        );
    }
}
