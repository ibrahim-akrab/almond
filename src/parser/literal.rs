//! Parsing for JS literals.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::{branch::alt, bytes::complete::*, combinator::*, number::complete::*, IResult};
use nom_locate::position;

pub fn parse_literal(s: Span) -> IResult<Span, Node> {
    ws0(alt((null_lit, bool_lit, array_lit)))(s)
}

pub fn null_lit(s: Span) -> IResult<Span, Node> {
    let (s, start) = position(s)?;
    let (s, _) = tag("null")(s)?;
    let (s, end) = position(s)?;
    Ok((s, LiteralValue::Null.into_node_kind().with_pos(start, end)))
}

pub fn bool_lit(s: Span) -> IResult<Span, Node> {
    let (s, start) = position(s)?;
    let (s, val) = alt((
        value(false, pair(tag("false"), not(identifier_continue))),
        value(true, pair(tag("true"), not(identifier_continue))),
    ))(s)?;
    let (s, end) = position(s)?;
    Ok((
        s,
        LiteralValue::Boolean(val)
            .into_node_kind()
            .with_pos(start, end),
    ))
}

pub fn numeric_lit(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;

    let (s, sign) =
        ws0(opt(char('-')))(s).map(|(s, minus)| (s, if minus.is_some() { -1 } else { 1 }))?;

    let (s, num) =
        alt((octal_int_lit, hex_int_lit, decimal_lit))(s).map(|(s, num)| (s, sign as f64 * num))?;
    let (s, end) = position(s)?;
    Ok((
        s,
        LiteralValue::Number(num)
            .into_node_kind()
            .with_pos(start, end),
    ))
}

fn decimal_lit(s: Span) -> ParseResult<f64> {
    recognize_float(s).map(|(s, float_str)| (s, float_str.parse::<f64>().unwrap()))
}

fn octal_int_lit(s: Span) -> ParseResult<f64> {
    let (s, octal_str) = preceded(char('0'), oct_digit1)(s)?;
    Ok((
        s,
        i64::from_str_radix(octal_str.fragment(), 8).unwrap() as f64,
    ))
}

fn hex_int_lit(s: Span) -> ParseResult<f64> {
    let (s, hex_str) = preceded(alt((tag("0x"), tag("0X"))), hex_digit1)(s)?;
    Ok((
        s,
        i64::from_str_radix(hex_str.fragment(), 16).unwrap() as f64,
    ))
}

pub fn array_lit(s: Span) -> IResult<Span, Node> {
    let (s, start) = position(s)?;
    let (s, expr_list) = delimited(terminated(char('['), spaces0), parse_expr_list, char(']'))(s)?;
    let (s, end) = position(s)?;
    spaces0(s)?;
    Ok((
        s,
        NodeKind::ArrayExpression {
            elements: expr_list,
        }
        .with_pos(start, end),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bool_lit() {
        bool_lit("true".into()).unwrap();
        bool_lit("false".into()).unwrap();
    }

    #[test]
    fn test_bool_lit_with_trailing_whitespace() {
        bool_lit("true ".into()).unwrap();
        bool_lit("false ".into()).unwrap();
    }

    #[test]
    fn test_bad_bool_lit() {
        bool_lit("bad".into()).unwrap_err();
        bool_lit("badfalse".into()).unwrap_err();
        bool_lit("falsebad".into()).unwrap_err();
    }

    #[test]
    fn test_numeric_lit() {
        assert_eq!(
            numeric_lit("123".into()).unwrap().1.kind,
            LiteralValue::Number(123.0).into_node_kind()
        );
        assert_eq!(
            numeric_lit("0123".into()).unwrap().1.kind,
            LiteralValue::Number(83.0).into_node_kind() // in octal
        );
        assert_eq!(
            numeric_lit("0x123".into()).unwrap().1.kind,
            LiteralValue::Number(291.0).into_node_kind() // in hex
        );
        assert_eq!(
            numeric_lit("0".into()).unwrap().1.kind,
            LiteralValue::Number(0.0).into_node_kind()
        );
        assert_eq!(
            numeric_lit("-123".into()).unwrap().1.kind,
            LiteralValue::Number(-123.0).into_node_kind()
        );
        assert_eq!(
            numeric_lit("-0".into()).unwrap().1.kind,
            LiteralValue::Number(-0.0).into_node_kind()
        );
        assert_eq!(
            numeric_lit("-0123".into()).unwrap().1.kind,
            LiteralValue::Number(-83.0).into_node_kind() // in octal
        );
        assert_eq!(
            numeric_lit("-0x123".into()).unwrap().1.kind,
            LiteralValue::Number(-291.0).into_node_kind() // in octal
        );
        assert_eq!(
            numeric_lit("- 123".into()).unwrap().1.kind,
            LiteralValue::Number(-123.0).into_node_kind()
        );
    }

    #[test]
    fn test_array_lit() {
        array_lit("[]".into()).unwrap();
        array_lit("[true, true]".into()).unwrap();
    }

    #[test]
    fn test_array_lit_with_whitespace() {
        array_lit("[ true, true ]\n".into()).unwrap();
        array_lit("[   true,\n true\t ]\t".into()).unwrap();
    }

    #[test]
    fn test_array_lit_with_trailing_comma() {
        array_lit("[ true, true, ]\n".into()).unwrap();
        array_lit("[   true,\n true,\t ]\t".into()).unwrap();
    }
}
