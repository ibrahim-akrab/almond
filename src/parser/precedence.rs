//! Precedence lookup table for various JS operators
//! Operator precedence parsing use Pratt parsing instead of LR because it is not possible with parser combinatorics.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::{branch::alt, bytes::complete::*, combinator::*};

#[derive(Debug, Copy, Clone)]
pub struct BindingPower(pub i32, pub i32);

#[derive(Debug, Copy, Clone)]
pub enum InfixOperator {
    Binary(BinaryOperator),
    Logical(LogicalOperator),
    Assignment(AssignmentOperator),
    /// Should be transformed into a `MemberExpression` with `computed = false`.
    DotOperator,
}
impl From<BinaryOperator> for InfixOperator {
    fn from(op: BinaryOperator) -> Self {
        InfixOperator::Binary(op)
    }
}
impl From<LogicalOperator> for InfixOperator {
    fn from(op: LogicalOperator) -> Self {
        InfixOperator::Logical(op)
    }
}
impl From<AssignmentOperator> for InfixOperator {
    fn from(op: AssignmentOperator) -> Self {
        InfixOperator::Assignment(op)
    }
}

/// Parses a binary (infix) operator.
/// Returns a tuple containing the operator and the operator binding power.
/// Refer to [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence) for details on JS operator precedence.
/// # Binding power
/// The binding power of the operator is based on the Mozilla documentation with some modifications for associativity.  
/// The lowest binding power for an operator is 1. Precedence 0 is to accept any expression.
///
/// The minimum of the left and right binding powers is always an odd number.
/// The maximum of the left and right binding powers is the double of the precedence on the Mozilla documentation page.
/// # Example
/// * Addition - Addition has a precedence of `14` and is left associative. **Binding power**: `(27, 28)`.
/// * Assignment - Assignment has a precedence of `3` and is right associative. **Binding power**: `(6, 5)`.
///
/// Returns `Err` if cannot parse a valid binary operator.
pub fn parse_infix_operator(s: Span) -> ParseResult<(InfixOperator, BindingPower)> {
    ws0(alt((
        // Assignment
        alt((
            value(
                (AssignmentOperator::Equals.into(), BindingPower(6, 5)),
                tag("="),
            ),
            value(
                (AssignmentOperator::PlusEquals.into(), BindingPower(6, 5)),
                tag("+="),
            ),
            value(
                (AssignmentOperator::MinusEquals.into(), BindingPower(6, 5)),
                tag("-="),
            ),
            value(
                (
                    AssignmentOperator::AsteriskEquals.into(),
                    BindingPower(6, 5),
                ),
                tag("*="),
            ),
            value(
                (AssignmentOperator::SlashEquals.into(), BindingPower(6, 5)),
                tag("/="),
            ),
            value(
                (AssignmentOperator::PercentEquals.into(), BindingPower(6, 5)),
                tag("%="),
            ),
            value(
                (
                    AssignmentOperator::ZeroFillLeftShiftEquals.into(),
                    BindingPower(6, 5),
                ),
                tag("<<="),
            ),
            value(
                (
                    AssignmentOperator::SignedRightShiftEquals.into(),
                    BindingPower(6, 5),
                ),
                tag(">>="),
            ),
            value(
                (
                    AssignmentOperator::ZeroFillRightShiftEquals.into(),
                    BindingPower(6, 5),
                ),
                tag(">>>="),
            ),
            value(
                (
                    AssignmentOperator::BitwiseAndEquals.into(),
                    BindingPower(6, 5),
                ),
                tag("&="),
            ),
            value(
                (
                    AssignmentOperator::BitwiseXorEquals.into(),
                    BindingPower(6, 5),
                ),
                tag("^="),
            ),
            value(
                (
                    AssignmentOperator::BitwiseOrEquals.into(),
                    BindingPower(6, 5),
                ),
                tag("|="),
            ),
        )),
        // Logical
        alt((
            value(
                (LogicalOperator::LogicalOr.into(), BindingPower(11, 12)),
                tag("||"),
            ),
            value(
                (LogicalOperator::LogicalAnd.into(), BindingPower(13, 14)),
                tag("&&"),
            ),
        )),
        // Bitwise
        value(
            (BinaryOperator::BitwiseOr.into(), BindingPower(15, 16)),
            tag("|"),
        ),
        value(
            (BinaryOperator::BitwiseXor.into(), BindingPower(17, 18)),
            tag("^"),
        ),
        value(
            (BinaryOperator::BitwiseAnd.into(), BindingPower(18, 20)),
            tag("&"),
        ),
        // Equality
        value(
            (BinaryOperator::EqualsEquals.into(), BindingPower(18, 20)),
            tag("=="),
        ),
        value(
            (BinaryOperator::NotEquals.into(), BindingPower(18, 20)),
            tag("!="),
        ),
        value(
            (BinaryOperator::TripleEquals.into(), BindingPower(18, 20)),
            tag("==="),
        ),
        value(
            (BinaryOperator::TripleNotEquals.into(), BindingPower(18, 20)),
            tag("!=="),
        ),
        // Additive
        value(
            (BinaryOperator::Plus.into(), BindingPower(27, 28)),
            tag("+"),
        ),
        value(
            (BinaryOperator::Minus.into(), BindingPower(27, 28)),
            tag("-"),
        ),
        // Multiplicative
        value(
            (BinaryOperator::Asterisk.into(), BindingPower(29, 30)),
            tag("*"),
        ),
        value(
            (BinaryOperator::Slash.into(), BindingPower(29, 30)),
            tag("/"),
        ),
        value(
            (BinaryOperator::Percent.into(), BindingPower(29, 30)),
            tag("%"),
        ),
        // Member Access
        value((InfixOperator::DotOperator, BindingPower(39, 40)), tag(".")),
        // TODO
    )))(s)
}

#[derive(Debug, Copy, Clone)]
pub enum PrefixOperator {
    Unary(UnaryOperator),
    Update(UpdateOperator),
}
impl From<UnaryOperator> for PrefixOperator {
    fn from(op: UnaryOperator) -> Self {
        PrefixOperator::Unary(op)
    }
}
impl From<UpdateOperator> for PrefixOperator {
    fn from(op: UpdateOperator) -> Self {
        PrefixOperator::Update(op)
    }
}

/// Parses a unary (prefix) operator.
/// Returns a tuple containing the operator and the operator binding power.
/// Refer to [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence) for details on JS operator precedence.
/// # Binding power
/// The binding power of the operator is based on the Mozilla documentation with some modifications for associativity.  
/// The lowest binding power for an operator is 1. Precedence 0 is to accept any expression.
///
/// The minimum of the left and right binding powers is always an odd number or `-1` (left for prefix).
/// The maximum of the left and right binding powers is the double of the precedence on the Mozilla documentation page.
/// # Example
/// * Unary Minus - Unary Minus has a precedence of 17 and is prefix. **Binding power**: `(-1, 34)`.
///
/// Returns `Err` if cannot parse a valid binary operator.
pub fn parse_prefix_operator(s: Span) -> ParseResult<(PrefixOperator, BindingPower)> {
    ws0(alt((
        // Update
        // Note: Update is matched first to prevent "+" and "-" to be matched first.
        value(
            (UpdateOperator::Increment.into(), BindingPower(-1, 34)),
            tag("++"),
        ),
        value(
            (UpdateOperator::Decrement.into(), BindingPower(-1, 34)),
            tag("--"),
        ),
        // Unary
        value(
            (UnaryOperator::Minus.into(), BindingPower(-1, 34)),
            tag("-"),
        ),
        value((UnaryOperator::Plus.into(), BindingPower(-1, 34)), tag("+")),
        value(
            (UnaryOperator::LogicalNot.into(), BindingPower(-1, 34)),
            tag("!"),
        ),
        value(
            (UnaryOperator::BitwiseNot.into(), BindingPower(-1, 34)),
            tag("~"),
        ),
        value(
            (UnaryOperator::Typeof.into(), BindingPower(-1, 34)),
            keyword_typeof,
        ),
        value(
            (UnaryOperator::Void.into(), BindingPower(-1, 34)),
            keyword_void,
        ),
        value(
            (UnaryOperator::Delete.into(), BindingPower(-1, 34)),
            keyword_delete,
        ),
    )))(s)
}
