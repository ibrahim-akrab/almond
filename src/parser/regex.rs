//! Parsing for JS regexp literals

use crate::ast::*;
use crate::parser::*;

fn regex_body(s: Span) -> ParseResult<String> {
    let escaped_regex_body = preceded(char('\\'), alt((char('\\'), char('/'))));
    fold_many0(
        alt((escaped_regex_body, none_of("/"))),
        format!(""),
        |mut body: String, regex_char| {
            body.push(regex_char);
            body
        },
    )(s)
}

fn regex_flags(s: Span) -> ParseResult<String> {
    fold_many0(
        identifier_continue,
        format!(""),
        |mut flag: String, flag_char| {
            flag.push(flag_char);
            flag
        },
    )(s)
}

pub fn regex_lit(s: Span) -> ParseResult<Node> {
    map(
        spanned(recognize(tuple((
            tag("/"),
            regex_body,
            tag("/"),
            regex_flags,
        )))),
        |(regex_str, start, end)| {
            LiteralValue::RegExp(regex_str.to_string())
                .into_node_kind()
                .with_pos(start, end)
        },
    )(s)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn test_regex_lit() {
        assert_json_snapshot!(parse_literal("/abc/g".into()).unwrap().1);
        assert_json_snapshot!(parse_literal(r#"/abc\\/g"#.into()).unwrap().1);
        assert_json_snapshot!(parse_literal(r#"/\\\//g"#.into()).unwrap().1);
    }
}
