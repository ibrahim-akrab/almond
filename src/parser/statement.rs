//! Parsing for JS statements.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::{branch::alt, bytes::complete::*, combinator::*};
use nom_locate::position;

pub fn parse_stmt(s: Span) -> ParseResult<Node> {
    alt((parse_block, parse_var_stmt, parse_empty_stmt))(s)
}

pub fn parse_block(s: Span) -> ParseResult<Node> {
    map(
        spanned(delimited(ws0(tag("{")), parse_stmt_list, tag("}"))),
        |(body, start, end)| NodeKind::BlockStatement { body }.with_pos(start, end),
    )(s)
}

pub fn parse_stmt_list(s: Span) -> ParseResult<Vec<Node>> {
    many0(parse_stmt)(s)
}

pub fn parse_var_stmt(s: Span) -> ParseResult<Node> {
    let initializer = preceded(ws0(tag("=")), parse_expr_no_seq);
    let var_declaration = map(
        tuple((position, pair(parse_identifier, opt(initializer)), position)),
        |(start, (id, init), end)| {
            NodeKind::VariableDeclarator {
                id: Box::new(id),
                init: Box::new(init),
            }
            .with_pos(start, end)
        },
    );
    let parse_declaration_list = separated_list1(ws0(tag(",")), ws0(var_declaration));

    let (s, start) = position(s)?;

    let (s, (declarations, end)) = spanned_end(delimited(
        ws1(keyword_var),
        parse_declaration_list,
        opt(semi),
    ))(s)?;

    Ok((
        s,
        NodeKind::VariableDeclaration {
            declarations,
            kind: VariableDeclarationKind::Var,
        }
        .with_pos(start, end),
    ))
}

pub fn parse_empty_stmt(s: Span) -> ParseResult<Node> {
    map(spanned(tag(";")), |(_, start, end)| {
        NodeKind::EmptyStatement.with_pos(start, end)
    })(s)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn test_var_stmt() {
        assert_json_snapshot!(parse_stmt("var x = 1;".into()).unwrap().1);
        assert_json_snapshot!(parse_stmt("var x, y = 1;".into()).unwrap().1);
        assert_json_snapshot!(parse_stmt("var x = 1, y;".into()).unwrap().1);
        assert_json_snapshot!(parse_stmt("var x".into()).unwrap().1); // auto insert semi
        assert_json_snapshot!(parse_stmt("var x\t".into()).unwrap().1);
    }

    #[test]
    fn test_empty_stmt() {
        assert_json_snapshot!(parse_stmt(";".into()).unwrap().1);
        assert_json_snapshot!(parse_stmt(";\t".into()).unwrap().1);
    }
}
