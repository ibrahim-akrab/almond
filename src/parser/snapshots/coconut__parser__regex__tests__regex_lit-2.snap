---
source: src/parser/regex.rs
expression: "parse_literal(r#\"/abc\\\\/g\"#.into()).unwrap().1"
---
{
  "type": "Literal",
  "value": "/abc\\\\/g",
  "start": 0,
  "end": 8
}
